comps/dataprep/neo4j/llama_index/extract_graph_neo4j.py

```diff
@@ -0,0 +1,590 @@
+            ),
+            ChatMessage(role="user", content=text),
+        ]
+        if OPENAI_API_KEY:
+            response = OpenAI().chat(messages)
+        else:
+            response = self.llm.chat(messages)
+
+        clean_response = re.sub(r"^assistant:\s*", "", str(response)).strip()
+        return clean_response
+
+    def build_communities(self):
+        """Builds communities from the graph and summarizes them."""
+        nx_graph = self._create_nx_graph()
+        community_hierarchical_clusters = hierarchical_leiden(nx_graph, max_cluster_size=self.max_cluster_size)
+        self.entity_info, community_info = self._collect_community_info(nx_graph, community_hierarchical_clusters)
+        self._summarize_communities(community_info)
+
+    def _create_nx_graph(self):
+        """Converts internal graph representation to NetworkX graph."""
+        nx_graph = nx.Graph()
+        triplets = self.get_triplets()
+        for entity1, relation, entity2 in triplets:
+            nx_graph.add_node(entity1.name)
+            nx_graph.add_node(entity2.name)
+            nx_graph.add_edge(
+                relation.source_id,
+                relation.target_id,
+                relationship=relation.label,
+                description=relation.properties["relationship_description"],
+            )
+        return nx_graph
+
+    def _collect_community_info(self, nx_graph, clusters):
+        """Collect information for each node based on their community,
+        allowing entities to belong to multiple clusters."""
+        entity_info = defaultdict(set)
+        community_info = defaultdict(list)
+
+        for item in clusters:
+            node = item.node
+            cluster_id = item.cluster
+
+            # Update entity_info
+            entity_info[node].add(cluster_id)
+
+            for neighbor in nx_graph.neighbors(node):
+                edge_data = nx_graph.get_edge_data(node, neighbor)
+                if edge_data:
+                    detail = f"{node} -> {neighbor} -> {edge_data['relationship']} -> {edge_data['description']}"
+                    community_info[cluster_id].append(detail)
+
+        # Convert sets to lists for easier serialization if needed
+        entity_info = {k: list(v) for k, v in entity_info.items()}
+
+        return dict(entity_info), dict(community_info)
+
+    def _summarize_communities(self, community_info):
+        """Generate and store summaries for each community."""
+        for community_id, details in community_info.items():
+            details_text = "\n".join(details) + "."  # Ensure it ends with a period
+            self.community_summary[community_id] = self.generate_community_summary(details_text)
+
+            # To store summaries in neo4j
+            # summary = self.generate_community_summary(details_text)
+            # self.community_summary[
+            #     community_id
+            # ] = self.store_community_summary_in_neo4j(community_id, summary)
```

Comment:
Do we still need this section?

Reply:
Although not used I would like to keep it, summaries are currently kept in memory in dictionary because is more efficient this way but depending on the scenario user may want to push those back to the DB and i want to make it easy to change the functionality since i already wrote the store to db method. is not a lot of lines.. i hope is OK :)